// @circle-fin/circle-sdk/dist/generated/models/ach-verification-error-codes.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AchVerificationErrorCodes = void 0;
/**
 * Indicates the failure reason of the ACH account. Only present on failed accounts. Possible values are [bank_account_authorization_expired, bank_account_error, bank_account_ineligible, bank_account_not_found, bank_account_unauthorized, unsupported_routing_number, verification_failed].
 * @export
 * @enum {string}
 */
exports.AchVerificationErrorCodes = {
    BankAccountAuthorizationExpired: "bank_account_authorization_expired",
    BankAccountError: "bank_account_error",
    BankAccountIneligible: "bank_account_ineligible",
    BankAccountNotFound: "bank_account_not_found",
    BankAccountUnauthorized: "bank_account_unauthorized",
    UnsupportedRoutingNumber: "unsupported_routing_number",
    VerificationFailed: "verification_failed"
};
