// @circle-fin/circle-sdk/dist/generated/apis/achapi.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ACHApi = exports.ACHApiFactory = exports.ACHApiFp = exports.ACHApiAxiosParamCreator = void 0;
const axios_1 = __importDefault(require("axios"));
// Some imports not used depending on template conditions
// @ts-ignore
const common_1 = require("../common");
// @ts-ignore
const base_1 = require("../base");
/**
 * ACHApi - axios parameter creator
 * @export
 */
const ACHApiAxiosParamCreator = function (configuration) {
    return {
        /**
         *
         * @summary Create a ACH bank account
         * @param {AchCreationRequest} [achCreationRequest]
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createAchAccount: (achCreationRequest, options = {}) => __awaiter(this, void 0, void 0, function* () {
            const localVarPath = `/v1/banks/ach`;
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: "POST" }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication bearerAuth required
            // http bearer authentication required
            yield (0, common_1.setBearerAuthToObject)(localVarHeaderParameter, configuration);
            localVarHeaderParameter["Content-Type"] = "application/json";
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            localVarRequestOptions.data = (0, common_1.serializeDataIfNeeded)(achCreationRequest, localVarRequestOptions, configuration);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions
            };
        }),
        /**
         * In the sandbox environment, create a mock ACH account and retrieve a processor token that can be used to link an ACH account.
         * @summary Create a mock ACH bank account
         * @param {MockAchAccountCreationRequest} [mockAchAccountCreationRequest]
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createMockAchAccount: (mockAchAccountCreationRequest, options = {}) => __awaiter(this, void 0, void 0, function* () {
            const localVarPath = `/v1/mocks/ach/accounts`;
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: "POST" }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication bearerAuth required
            // http bearer authentication required
            yield (0, common_1.setBearerAuthToObject)(localVarHeaderParameter, configuration);
            localVarHeaderParameter["Content-Type"] = "application/json";
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            localVarRequestOptions.data = (0, common_1.serializeDataIfNeeded)(mockAchAccountCreationRequest, localVarRequestOptions, configuration);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions
            };
        }),
        /**
         *
         * @summary Get a ACH bank account
         * @param {string} id Universally unique identifier (UUID v4) of a resource.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getAchAccount: (id, options = {}) => __awaiter(this, void 0, void 0, function* () {
            // verify required parameter 'id' is not null or undefined
            (0, common_1.assertParamExists)("getAchAccount", "id", id);
            const localVarPath = `/v1/banks/ach/{id}`.replace(`{${"id"}}`, encodeURIComponent(String(id)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: "GET" }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication bearerAuth required
            // http bearer authentication required
            yield (0, common_1.setBearerAuthToObject)(localVarHeaderParameter, configuration);
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions
            };
        })
    };
};
exports.ACHApiAxiosParamCreator = ACHApiAxiosParamCreator;
/**
 * ACHApi - functional programming interface
 * @export
 */
const ACHApiFp = function (configuration) {
    const localVarAxiosParamCreator = (0, exports.ACHApiAxiosParamCreator)(configuration);
    return {
        /**
         *
         * @summary Create a ACH bank account
         * @param {AchCreationRequest} [achCreationRequest]
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createAchAccount(achCreationRequest, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.createAchAccount(achCreationRequest, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
        /**
         * In the sandbox environment, create a mock ACH account and retrieve a processor token that can be used to link an ACH account.
         * @summary Create a mock ACH bank account
         * @param {MockAchAccountCreationRequest} [mockAchAccountCreationRequest]
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createMockAchAccount(mockAchAccountCreationRequest, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.createMockAchAccount(mockAchAccountCreationRequest, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
        /**
         *
         * @summary Get a ACH bank account
         * @param {string} id Universally unique identifier (UUID v4) of a resource.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getAchAccount(id, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.getAchAccount(id, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        }
    };
};
exports.ACHApiFp = ACHApiFp;
/**
 * ACHApi - factory interface
 * @export
 */
const ACHApiFactory = function (configuration, basePath, axios) {
    const localVarFp = (0, exports.ACHApiFp)(configuration);
    return {
        /**
         *
         * @summary Create a ACH bank account
         * @param {AchCreationRequest} [achCreationRequest]
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createAchAccount(achCreationRequest, options) {
            return localVarFp
                .createAchAccount(achCreationRequest, options)
                .then((request) => request(axios, basePath));
        },
        /**
         * In the sandbox environment, create a mock ACH account and retrieve a processor token that can be used to link an ACH account.
         * @summary Create a mock ACH bank account
         * @param {MockAchAccountCreationRequest} [mockAchAccountCreationRequest]
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createMockAchAccount(mockAchAccountCreationRequest, options) {
            return localVarFp
                .createMockAchAccount(mockAchAccountCreationRequest, options)
                .then((request) => request(axios, basePath));
        },
        /**
         *
         * @summary Get a ACH bank account
         * @param {string} id Universally unique identifier (UUID v4) of a resource.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getAchAccount(id, options) {
            return localVarFp
                .getAchAccount(id, options)
                .then((request) => request(axios, basePath));
        }
    };
};
exports.ACHApiFactory = ACHApiFactory;
/**
 * ACHApi - object-oriented interface
 * @export
 * @class ACHApi
 * @extends {BaseAPI}
 */
class ACHApi extends base_1.BaseAPI {
    /**
     *
     * @summary Create a ACH bank account
     * @param {AchCreationRequest} [achCreationRequest]
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof ACHApi
     */
    createAchAccount(achCreationRequest, options) {
        return (0, exports.ACHApiFp)(this.configuration)
            .createAchAccount(achCreationRequest, options)
            .then((request) => request(this.axios, this.basePath));
    }
    /**
     * In the sandbox environment, create a mock ACH account and retrieve a processor token that can be used to link an ACH account.
     * @summary Create a mock ACH bank account
     * @param {MockAchAccountCreationRequest} [mockAchAccountCreationRequest]
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof ACHApi
     */
    createMockAchAccount(mockAchAccountCreationRequest, options) {
        return (0, exports.ACHApiFp)(this.configuration)
            .createMockAchAccount(mockAchAccountCreationRequest, options)
            .then((request) => request(this.axios, this.basePath));
    }
    /**
     *
     * @summary Get a ACH bank account
     * @param {string} id Universally unique identifier (UUID v4) of a resource.
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof ACHApi
     */
    getAchAccount(id, options) {
        return (0, exports.ACHApiFp)(this.configuration)
            .getAchAccount(id, options)
            .then((request) => request(this.axios, this.basePath));
    }
}
exports.ACHApi = ACHApi;
