// @circle-fin/circle-sdk/dist/generated/apis/sepaapi.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SEPAApi = exports.SEPAApiFactory = exports.SEPAApiFp = exports.SEPAApiAxiosParamCreator = void 0;
const axios_1 = __importDefault(require("axios"));
// Some imports not used depending on template conditions
// @ts-ignore
const common_1 = require("../common");
// @ts-ignore
const base_1 = require("../base");
/**
 * SEPAApi - axios parameter creator
 * @export
 */
const SEPAApiAxiosParamCreator = function (configuration) {
    return {
        /**
         *
         * @summary Create a SEPA bank account (BETA)
         * @param {SepaCreationRequest} [sepaCreationRequest]
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createSepaAccount: (sepaCreationRequest, options = {}) => __awaiter(this, void 0, void 0, function* () {
            const localVarPath = `/v1/banks/sepa`;
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: "POST" }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication bearerAuth required
            // http bearer authentication required
            yield (0, common_1.setBearerAuthToObject)(localVarHeaderParameter, configuration);
            localVarHeaderParameter["Content-Type"] = "application/json";
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            localVarRequestOptions.data = (0, common_1.serializeDataIfNeeded)(sepaCreationRequest, localVarRequestOptions, configuration);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions
            };
        }),
        /**
         *
         * @summary Get a SEPA bank account (BETA)
         * @param {string} id Universally unique identifier (UUID v4) of a resource.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getSepaAccount: (id, options = {}) => __awaiter(this, void 0, void 0, function* () {
            // verify required parameter 'id' is not null or undefined
            (0, common_1.assertParamExists)("getSepaAccount", "id", id);
            const localVarPath = `/v1/banks/sepa/{id}`.replace(`{${"id"}}`, encodeURIComponent(String(id)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: "GET" }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication bearerAuth required
            // http bearer authentication required
            yield (0, common_1.setBearerAuthToObject)(localVarHeaderParameter, configuration);
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions
            };
        }),
        /**
         * Get the SEPA transfer instructions into the Circle bank account given your bank account id (BETA).
         * @summary Get SEPA instructions (BETA)
         * @param {string} id Universally unique identifier (UUID v4) of a resource.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getSepaAccountInstructions: (id, options = {}) => __awaiter(this, void 0, void 0, function* () {
            // verify required parameter 'id' is not null or undefined
            (0, common_1.assertParamExists)("getSepaAccountInstructions", "id", id);
            const localVarPath = `/v1/banks/sepa/{id}/instructions`.replace(`{${"id"}}`, encodeURIComponent(String(id)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: "GET" }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication bearerAuth required
            // http bearer authentication required
            yield (0, common_1.setBearerAuthToObject)(localVarHeaderParameter, configuration);
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions
            };
        })
    };
};
exports.SEPAApiAxiosParamCreator = SEPAApiAxiosParamCreator;
/**
 * SEPAApi - functional programming interface
 * @export
 */
const SEPAApiFp = function (configuration) {
    const localVarAxiosParamCreator = (0, exports.SEPAApiAxiosParamCreator)(configuration);
    return {
        /**
         *
         * @summary Create a SEPA bank account (BETA)
         * @param {SepaCreationRequest} [sepaCreationRequest]
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createSepaAccount(sepaCreationRequest, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.createSepaAccount(sepaCreationRequest, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
        /**
         *
         * @summary Get a SEPA bank account (BETA)
         * @param {string} id Universally unique identifier (UUID v4) of a resource.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getSepaAccount(id, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.getSepaAccount(id, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
        /**
         * Get the SEPA transfer instructions into the Circle bank account given your bank account id (BETA).
         * @summary Get SEPA instructions (BETA)
         * @param {string} id Universally unique identifier (UUID v4) of a resource.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getSepaAccountInstructions(id, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.getSepaAccountInstructions(id, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        }
    };
};
exports.SEPAApiFp = SEPAApiFp;
/**
 * SEPAApi - factory interface
 * @export
 */
const SEPAApiFactory = function (configuration, basePath, axios) {
    const localVarFp = (0, exports.SEPAApiFp)(configuration);
    return {
        /**
         *
         * @summary Create a SEPA bank account (BETA)
         * @param {SepaCreationRequest} [sepaCreationRequest]
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createSepaAccount(sepaCreationRequest, options) {
            return localVarFp
                .createSepaAccount(sepaCreationRequest, options)
                .then((request) => request(axios, basePath));
        },
        /**
         *
         * @summary Get a SEPA bank account (BETA)
         * @param {string} id Universally unique identifier (UUID v4) of a resource.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getSepaAccount(id, options) {
            return localVarFp
                .getSepaAccount(id, options)
                .then((request) => request(axios, basePath));
        },
        /**
         * Get the SEPA transfer instructions into the Circle bank account given your bank account id (BETA).
         * @summary Get SEPA instructions (BETA)
         * @param {string} id Universally unique identifier (UUID v4) of a resource.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getSepaAccountInstructions(id, options) {
            return localVarFp
                .getSepaAccountInstructions(id, options)
                .then((request) => request(axios, basePath));
        }
    };
};
exports.SEPAApiFactory = SEPAApiFactory;
/**
 * SEPAApi - object-oriented interface
 * @export
 * @class SEPAApi
 * @extends {BaseAPI}
 */
class SEPAApi extends base_1.BaseAPI {
    /**
     *
     * @summary Create a SEPA bank account (BETA)
     * @param {SepaCreationRequest} [sepaCreationRequest]
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof SEPAApi
     */
    createSepaAccount(sepaCreationRequest, options) {
        return (0, exports.SEPAApiFp)(this.configuration)
            .createSepaAccount(sepaCreationRequest, options)
            .then((request) => request(this.axios, this.basePath));
    }
    /**
     *
     * @summary Get a SEPA bank account (BETA)
     * @param {string} id Universally unique identifier (UUID v4) of a resource.
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof SEPAApi
     */
    getSepaAccount(id, options) {
        return (0, exports.SEPAApiFp)(this.configuration)
            .getSepaAccount(id, options)
            .then((request) => request(this.axios, this.basePath));
    }
    /**
     * Get the SEPA transfer instructions into the Circle bank account given your bank account id (BETA).
     * @summary Get SEPA instructions (BETA)
     * @param {string} id Universally unique identifier (UUID v4) of a resource.
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof SEPAApi
     */
    getSepaAccountInstructions(id, options) {
        return (0, exports.SEPAApiFp)(this.configuration)
            .getSepaAccountInstructions(id, options)
            .then((request) => request(this.axios, this.basePath));
    }
}
exports.SEPAApi = SEPAApi;
