// @circle-fin/circle-sdk/dist/generated/index.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Circle = void 0;
const api_1 = require("./api");
const configuration_1 = require("./configuration");
__exportStar(require("./models"), exports);
class Circle {
    constructor(apiKey, baseUrl) {
        const configuration = new configuration_1.Configuration({ basePath: baseUrl, apiKey });
        this.ach = new api_1.ACHApi(configuration);
        this.addresses = new api_1.AddressesApi(configuration);
        this.balances = new api_1.BalancesApi(configuration);
        this.cards = new api_1.CardsApi(configuration);
        this.channels = new api_1.ChannelsApi(configuration);
        this.chargebacks = new api_1.ChargebacksApi(configuration);
        this.checkoutSessions = new api_1.CheckoutSessionsApi(configuration);
        this.cryptoAddressBook = new api_1.CryptoAddressBookApi(configuration);
        this.cryptoExchangeRates = new api_1.CryptoExchangeRatesApi(configuration);
        this.cryptoPaymentIntents = new api_1.CryptoPaymentIntentsApi(configuration);
        this.deposits = new api_1.DepositsApi(configuration);
        this.encryption = new api_1.EncryptionApi(configuration);
        this.health = new api_1.HealthApi(configuration);
        this.management = new api_1.ManagementApi(configuration);
        this.paymentTokens = new api_1.PaymentTokensApi(configuration);
        this.payments = new api_1.PaymentsApi(configuration);
        this.payouts = new api_1.PayoutsApi(configuration);
        this.returns = new api_1.ReturnsApi(configuration);
        this.reversals = new api_1.ReversalsApi(configuration);
        this.sen = new api_1.SENApi(configuration);
        this.sepa = new api_1.SEPAApi(configuration);
        this.settlements = new api_1.SettlementsApi(configuration);
        this.signet = new api_1.SignetApi(configuration);
        this.stablecoins = new api_1.StablecoinsApi(configuration);
        this.subscriptions = new api_1.SubscriptionsApi(configuration);
        this.transfers = new api_1.TransfersApi(configuration);
        this.wallets = new api_1.WalletsApi(configuration);
        this.wires = new api_1.WiresApi(configuration);
    }
}
exports.Circle = Circle;
